### PR TITLE
Recipe edit

### PR DESCRIPTION
**Pull request by jpalmieri**
I think the heavy lifting of this branch is done and ready for review.

I'm going to add the dynamic fields to the edit view, but it shouldn't be that big of a deal.

### CODE REVIEW

File: app/models/recipe.rb [hunk 1]
```diff
@@ -46,4 +48,14 @@ def total_grain_weight
...
 
+  def reject_blank_grains_and_hops
+    [grains, hops].each do |grain_or_hop|
+      grain_or_hop.each do |ingredient|
+        if ingredient.weight.blank? && ingredient.name.blank?
+          ingredient.mark_for_destruction
+        end
+      end
+    end
+  end
```
Comment (jpalmieri) [inline app/models/recipe.rb:71]: http://stackoverflow.com/questions/10943013/destroy-on-blank-nested-attribute

File: app/assets/javascripts/recipes.js [hunk 2]
```diff
@@ -11,15 +9,22 @@ $(document).ready(function(){
...
     if(index > 0 && inputsAreBlank){
       $(grainElement).hide();
-    } else if (index > 0) {
-      $(grainElement).prev(".grain").find("button.add-grain").hide();
-      // Only show button for last non-blank input
     }
-   });
+  });
+
+  var show_last_add_grain_button = function(){
```
Comment (eliotsykes) [inline app/assets/javascripts/recipes.js]: More conventional to use camelCase for naming JS variables.

File: app/assets/javascripts/recipes.js [hunk 3]
```diff
@@ -11,15 +9,22 @@ $(document).ready(function(){
...
+        return $(this);
+      }
+    }).last().find('button.add-grain').show();
+  };
+  show_last_add_grain_button();
 
   $("button.add-grain").on("click", function(event) {
     event.preventDefault();
-    $(this).parent(".grain").next(".grain").show();
+    $(this).parent(".grain").nextAll(".grain").first().show();
```
Comment (eliotsykes) [inline app/assets/javascripts/recipes.js]: Try to avoid repeated jQuery element $(wrapping) where an element can be cached as a variable. In this case:

``` javascript
$(this).parent(".grain").nextAll(".grain").first().show();
$(this).hide();
```

becomes:

``` javascript
var addButton = $(this);
addButton.parent(".grain").nextAll(".grain").first().show();
addButton.hide();
```

File: app/assets/javascripts/recipes.js [hunk 4]
```diff
@@ -11,15 +9,22 @@ $(document).ready(function(){
...
-    } else if (index > 0) {
-      $(grainElement).prev(".grain").find("button.add-grain").hide();
-      // Only show button for last non-blank input
     }
-   });
+  });
+
+  var show_last_add_grain_button = function(){
+    $('.grain').filter( function(){ 
+      if($(this).css('display') == "block") {
```
Comment (eliotsykes) [inline app/assets/javascripts/recipes.js]: Repeated wrapping of $(this) could be cached, see comment below: https://github.com/jpalmieri/brewery/pull/11/files#r24754377

File: app/assets/javascripts/recipes.js (hunk 3 above)
Comment (eliotsykes) [inline app/assets/javascripts/recipes.js]: Consider only having one `<button class="add-grain">` and moving it after/outside the `.grain` divs, as it may reduce the amount of code needed here for managing button visibility. Admittedly it may not too, just a thought.

File: app/assets/javascripts/recipes.js (hunk 4 above)
Comment (eliotsykes) [inline app/assets/javascripts/recipes.js]: The jQuery `:visible` selector extension may be useful here to save the `display == block` check, and make the code a little more understandable and concise. http://api.jquery.com/visible-selector/

Note that `:visible` selector is a jQuery extension and not available in CSS.

Also note that `:visible` can have a performance cost. The jQuery docs (linked above) recommend using it after do the initial element selection with a `filter(":visible")` call. So instead of: 

``` javascript
$(".grain:visible")
```

prefer this:

``` javascript
$(".grain").filter(":visible")
```

It's less clear to me if the `:last-child` selector might also be useful here: http://api.jquery.com/last-child-selector/

File: app/controllers/recipes_controller.rb [hunk 5]
```diff
@@ -21,6 +21,36 @@ def create
     end
   end
 
+  def edit
+    if current_user.id == Recipe.find(params[:id]).user.id
```
Comment (eliotsykes) [inline app/controllers/recipes_controller.rb]: Consider using user object equality comparison here, without the id calls:

``` ruby
if current_user == Recipe.find(params[:id]).user
```

Comment (eliotsykes) [inline app/controllers/recipes_controller.rb]: It's possible here to reduce the number of db queries looking up on the recipes table without sacrificing security. For performance, it's good to consider reducing the number of db queries where it's **easily** done with a minor tweak. Currently line 25 and line 26 both lookup on the recipes table almost identically.

File: app/controllers/recipes_controller.rb [hunk 6]
```diff
@@ -21,6 +21,36 @@ def create
...
+    else
+      flash[:error] = "You are not authorized to edit this recipe."
+      redirect_to root_path
+    end
+  end
+
+  def update
+    if current_user.id == Recipe.find(params[:id]).user.id
+      @recipe = current_user.recipes.find(params[:id])
+      if @recipe.update_attributes(recipe_params)
```
Comment (eliotsykes) [inline app/controllers/recipes_controller.rb:38]: Indentation in this method needs work, its a little hard to follow at the moment

File: app/controllers/recipes_controller.rb [hunk 7]
```diff
@@ -21,6 +21,36 @@ def create
...
+        flash[:notice] = "Recipe updated."
+        redirect_to @recipe
+      else
+        error_list = "<ul>"
```
Comment (eliotsykes) [inline app/controllers/recipes_controller.rb]: Try to avoid building up HTML strings in the controller, and aim to keep HTML in the view templates and view helpers where possible.

The `@recipe.errors.full_messages` array can be accessed in the edit template to build up an HTML list. The `@recipe.valid?` method may also be useful.

File: app/models/recipe.rb (hunk 1 above)
Comment (eliotsykes) [inline app/models/recipe.rb:71]: Nice work on `mark_for_destruction` (I'd not encountered that, thanks for the tip).

It's starting to look like the `Grain`, `Hop`, and `Yeast` models might benefit from having their own `blank?` methods that wrap the `weight.blank? && name.blank?` etc. checks, e.g.

``` ruby
class Grain ...
  ...
  def blank?
    weight.blank? && name.blank?
  end
  ...
end
```

File: app/views/recipes/edit.html.haml [hunk 8]
```diff
@@ -0,0 +1,36 @@
+%h1 New Recipe
```
Comment (eliotsykes) [inline app/views/recipes/edit.html.haml]: **Edit** Recipe

File: spec/factories/recipe.rb [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
-    name "Name"
-    style "Style"
+    name "Super dark beer"
+    style "Black Ale"
     batch_size 2.5
 
     before(:create) do |recipe, yeast|
       yeasts_attributes = []
-      3.times do 
+      1.times do
```
Comment (eliotsykes) [inline spec/factories/recipe.rb]: `1.times` needed?

Comment (jpalmieri) [inline spec/factories/recipe.rb]: :smile: Ha

That code is redundant, so I'll remove it since it's redundant.

File: app/assets/javascripts/recipes.js (hunk 4 above)
Comment (jpalmieri) [inline app/assets/javascripts/recipes.js]: :+1: thanks

File: spec/factories/recipe.rb (hunk 9 above)
Comment (jpalmieri) [inline spec/factories/recipe.rb]: I fixed this on the master branch

**Comment by jpalmieri** [comment]
Do the additional changes look ok?

**Comment by jpalmieri** [comment]
I'm going to merge this to avoid extra weirdness when starting work on a new branch, but let me know if you have any more comments.
